presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -327,6 +331,69 @@ public static long truncate(@SqlType(StandardTypes.REAL) long num)
         return floatToRawIntBits((float) (Math.signum(numInFloat) * Math.floor(Math.abs(numInFloat))));
     }
 
+    @Description("truncate to double by dropping digits after decimal point")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double truncate(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(StandardTypes.BIGINT) long decimals)
+    {
+        if (Double.isNaN(num) || Double.isInfinite(num)) {
+            // compatible with truncate(DOUBLE)
+            return num;
+        }
+        if (decimals == 0) {
+            if (num >= 0) {
+                return Math.floor(num);
+            }
+            else {
+                return Math.ceil(num);
+            }
+        }
+
+        if (decimals < 0) {
+            if (decimals < -NUMBER_LENGTH) {
+                decimals = -NUMBER_LENGTH;
+            }
+        }
+        else {
+            if (decimals >= NUMBER_LENGTH) {
+                decimals = NUMBER_LENGTH;
+            }
+        }
+        return BigDecimal.valueOf(num).setScale((int) decimals, BigDecimal.ROUND_DOWN).doubleValue();
+    }
+
+    @Description("truncate to float by dropping digits after decimal point")
+    @ScalarFunction
+    @SqlType(StandardTypes.REAL)
+    public static long truncate(@SqlType(StandardTypes.REAL) long num, @SqlType(StandardTypes.BIGINT) long decimals)
+    {
+        float numBitsToFloats = intBitsToFloat((int) num);
+        if (Float.isNaN(numBitsToFloats) || Float.isInfinite(numBitsToFloats)) {
+            // compatible with truncate(REAL)
+            return num;
+        }
+        if (decimals == 0) {
+            if (numBitsToFloats >= 0) {
+                return floatToRawIntBits((float) Math.floor(numBitsToFloats));
+            }
+            else {
+                return floatToRawIntBits((float) Math.ceil(numBitsToFloats));
+            }
+        }
+
+        if (decimals < 0) {
+            if (decimals < -NUMBER_LENGTH) {
+                decimals = -NUMBER_LENGTH;
+            }
+        }
+        else {
+            if (decimals >= NUMBER_LENGTH) {
+                decimals = NUMBER_LENGTH;
+            }
+        }
+        return floatToRawIntBits(new BigDecimal(Float.toString(numBitsToFloats)).setScale((int) decimals, BigDecimal.ROUND_DOWN).floatValue());
```

Comment:
```suggestion
        return floatToRawIntBits(new BigDecimal(numBitsToFloats).setScale((int) decimals, BigDecimal.ROUND_DOWN).floatValue());
```

Any reason why this had to be converted to a string?  The test cases pass with the above.

Reply:
BigDecimal takes type of DOUBLE not FLOAT so use String as a safe cast value to return.